### PR TITLE
Implementation of the new binlog-server command

### PR DESCRIPTION
**Pull request by Fizic**
### Database name
mysql

# Pull request description

wal-g runs MySQL implementation, and when a replica is connected to it, downloads binlogs from the repository and gives them over the replication protocol.

Currently requires the use of the library version provided in this pull-request: https://github.com/go-mysql-org/go-mysql/pull/759

### CODE REVIEW

**Comment by lance6716** [comment]
Seems you can use go mod replace to let go-mysql use https://github.com/go-mysql-org/go-mysql/pull/759 and let the CI work :)

**Comment by ostinru** [comment]
Wow! Great beginning!
I thought about this approach after reading ‘MySQL PiTR like a rockstar’[1][2][3] by @lefred
Below is my RFC I did some time ago, I hope you will find bits of it useful.

---
## Proposal [stage 1]

add new command `wal-g binlog-server`  that

- start listening a port
- parse requests on incoming connections (basic auth?)
- send *all* binlogs since backup timestamp (like we are doing it now) (known bug: https://github.com/wal-g/wal-g/issues/1386)
- be able to stop sending events when reaching point in time

**Open questions/discussion:**

1. we should parse binlogs to find PiT. Is it fast to parse binlogs in golang? (anyway, mysqlbinlog also parses binlogs)
2. at first glance `github.com/go-mysql-org/go-mysql` lacks ready-to-use code for server message parsing

## Proposal [stage 2]

This stage aims to make binlog-server generic: unlink it from backup_name.

for `wal-g binlog-push` - we should save GTID Sets to binlog sentinel, so `binlog-server` will be able to find the first binlog to serve

- Q: is there a chance for race between instances of wal-g writing the sentinel file?

for `wal-g binlog-server` we should be able to parse mysqld's [COM_BINLOG_DUMP_GTID](https://github.com/mysql/mysql-server/blob/mysql-cluster-8.0.30/sql/rpl_source.cc#L936
) request (get GTIDSet from it) and find a set of binlog files to be sent to `mysqld`



**Open questions/discussion:**

1. Should we add ‘test’ command that will check that we have information about all required binlogs (to allow users to fallback to the old approach if we are not sure)

# Other considerations

1. We should replay the first binlog from exact position: https://github.com/wal-g/wal-g/issues/1386

---

### Links
[1] https://lefred.be/content/howto-make-mysql-point-in-time-recovery-faster/
[2] https://www.slideshare.net/lefred.descamps/fosdem-mysql-friends-devroomffebruary-2018-mysql-pointintime-recovery-like-a-rockstar
[3] https://www.youtube.com/watch?v=PinKYCfv1MM&t=1787s

File: go.mod [hunk 1]
```diff
@@ -62,6 +62,7 @@ require (
 	github.com/pkg/profile v1.6.0
 	github.com/prometheus/client_golang v1.12.1
 	github.com/prometheus/client_model v0.2.0
+	github.com/siddontang/go-log v0.0.0-20190221022429-1e957dd83bed
```
Comment (teem0n) [inline go.mod:65]: It seems that we don't need one more log. Let's remove this one

File: internal/databases/mysql/mysql.go [hunk 2]
```diff
@@ -205,6 +220,32 @@ type binlogHandler interface {
...
+	logsToFetch, err := getLogsAfterBinlog(logFolder, binlogName)
+	if err != nil {
+		return err
+	}
+	for _, logFile := range logsToFetch {
+		binlogName := utility.TrimFileExtension(logFile.GetName())
+		binlogPath := path.Join(dstDir, binlogName)
+		tracelog.InfoLogger.Printf("downloading %s into %s", binlogName, binlogPath)
+		if err = internal.DownloadFileTo(logFolder, binlogName, binlogPath); err != nil {
+			if errors.Is(err, afero.ErrFileExists) {
```
Comment (teem0n) [inline internal/databases/mysql/mysql.go:234]: Let's replace with `os.IsExist(err)`

File: internal/databases/mysql/binlog_server_handler.go [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+var (
+	untilTS      time.Time
+	lastSentGTID string
+)
+
+func addRotateEvent(s *replication.BinlogStreamer, pos mysql.Position) error {
+	// create rotate event
+	rotateBinlogEvent := replication.BinlogEvent{}
+	rotateBinlogEvent.RawData = make([]byte, 4+1+4+4+4+2+8)
```
Comment (teem0n) [inline internal/databases/mysql/binlog_server_handler.go:31]: This magic number should be placed in const var with some description and links in comment : )

File: internal/config.go [hunk 4]
```diff
@@ -107,13 +107,18 @@ const (
...
-	MysqlTakeBinlogsFromMaster = "WALG_MYSQL_TAKE_BINLOGS_FROM_MASTER"
-	MysqlCheckGTIDs            = "WALG_MYSQL_CHECK_GTIDS"
+	MysqlDatasourceNameSetting    = "WALG_MYSQL_DATASOURCE_NAME"
+	MysqlSslCaSetting             = "WALG_MYSQL_SSL_CA"
+	MysqlBinlogReplayCmd          = "WALG_MYSQL_BINLOG_REPLAY_COMMAND"
+	MysqlBinlogDstSetting         = "WALG_MYSQL_BINLOG_DST"
+	MysqlBackupPrepareCmd         = "WALG_MYSQL_BACKUP_PREPARE_COMMAND"
+	MysqlTakeBinlogsFromMaster    = "WALG_MYSQL_TAKE_BINLOGS_FROM_MASTER"
+	MysqlCheckGTIDs               = "WALG_MYSQL_CHECK_GTIDS"
+	MysqlBinlogGetExecutedGTIDCmd = "WALG_MYSQL_BINLOG_GET_EXECUTED_GTID_COMMAND"
```
Comment (teem0n) [inline internal/config.go:117]: This command needs description in .md file with example (like WALG_MYSQL_BINLOG_REPLAY_COMMAND)

File: internal/databases/mysql/binlog_server_handler.go [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+	// set data
+	binary.LittleEndian.PutUint64(rotateBinlogEvent.RawData[binlogEventPos:], uint64(pos.Pos))
+	rotateBinlogEvent.RawData = append(rotateBinlogEvent.RawData, []byte(pos.Name)...)
+
+	return s.AddEventToStreamer(&rotateBinlogEvent)
+}
+
+func waitReplicationIsDone() error {
+	for {
+		cmd, err := internal.GetCommandSetting(internal.MysqlBinlogGetExecutedGTIDCmd)
```
Comment (teem0n) [inline internal/databases/mysql/binlog_server_handler.go:54]: May be we can use getMySQLGTIDExecuted here?

File: internal/databases/mysql/binlog_server_handler.go [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+			return err
+		}
+		var out bytes.Buffer
+		cmd.Stdout = &out
+		err = cmd.Run()
+		if err != nil {
+			return err
+		}
+		executedGTIDSet := out.String()
+		if strings.Contains(executedGTIDSet, lastSentGTID) {
```
Comment (teem0n) [inline internal/databases/mysql/binlog_server_handler.go:65]: lastSentGTID should be read under some kind of lock, like sync.Mutex
Also, we can not use string.Contains here, we should parse lastSentGTID with gomysql.ParseGTIDSet

File: internal/databases/mysql/binlog_server_handler.go [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+
+	p := replication.NewBinlogParser()
+
+	f := func(e *replication.BinlogEvent) error {
+		if int64(e.Header.Timestamp) > untilTS.Unix() {
+			return nil
+		}
+		if e.Header.EventType == replication.GTID_EVENT {
+			u, _ := uuid.FromBytes(e.Event.(*replication.GTIDEvent).SID)
+			lastSentGTID = u.String() + ":1-" + strconv.Itoa(int(e.Event.(*replication.GTIDEvent).GNO))
```
Comment (teem0n) [inline internal/databases/mysql/binlog_server_handler.go:91]: Does we need "1-" here?
Can we just send SID:GNO here?

File: internal/databases/mysql/binlog_server_handler.go [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+		}
+		time.Sleep(1 * time.Second)
+	}
+	return nil
+}
+
+func sendEventsFromBinlogFiles(logFilesProvider *storage.ObjectProvider, pos mysql.Position, s *replication.BinlogStreamer) {
+	err := addRotateEvent(s, pos)
+	if err != nil {
+		ok := s.AddErrorToStreamer(err)
```
Comment (teem0n) [inline internal/databases/mysql/binlog_server_handler.go:77]: Let's put this logic in method like "HandleEventError"

File: internal/databases/mysql/binlog_server_handler.go [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+		resultSet, _ := mysql.BuildSimpleTextResultset([]string{"SERVER_ID"}, [][]interface{}{{"10000"}})
+		return &mysql.Result{Status: 34, Warnings: 0, InsertId: 0, AffectedRows: 0, Resultset: resultSet}, nil
+	case "SELECT @source_binlog_checksum":
+		resultSet, _ := mysql.BuildSimpleTextResultset([]string{"source_binlog_checksum"}, [][]interface{}{{"1"}})
+		return &mysql.Result{Status: 34, Warnings: 0, InsertId: 0, AffectedRows: 0, Resultset: resultSet}, nil
+	case "SELECT @@GLOBAL.GTID_MODE":
+		resultSet, _ := mysql.BuildSimpleTextResultset([]string{"GTID_MODE"}, [][]interface{}{{"ON"}})
+		return &mysql.Result{Status: 34, Warnings: 0, InsertId: 0, AffectedRows: 0, Resultset: resultSet}, nil
+	case "SELECT @@GLOBAL.SERVER_UUID":
+		resultSet, _ := mysql.BuildSimpleTextResultset([]string{"SERVER_UUID"}, [][]interface{}{{"123"}})
```
Comment (teem0n) [inline internal/databases/mysql/binlog_server_handler.go:199]: 123 should be read from config

File: internal/databases/mysql/binlog_server_handler.go [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+func (h Handler) HandleQuery(query string) (*mysql.Result, error) {
+	switch query {
+	case "SELECT @master_binlog_checksum":
+		resultSet, _ := mysql.BuildSimpleTextResultset([]string{"master_binlog_checksum"}, [][]interface{}{{"NONE"}})
+		return &mysql.Result{Status: 34, Warnings: 0, InsertId: 0, AffectedRows: 0, Resultset: resultSet}, nil
+	case "SHOW GLOBAL VARIABLES LIKE 'BINLOG_CHECKSUM'":
+		resultSet, _ := mysql.BuildSimpleTextResultset([]string{"BINLOG_CHECKSUM"}, [][]interface{}{{"NONE"}})
+		return &mysql.Result{Status: 34, Warnings: 0, InsertId: 0, AffectedRows: 0, Resultset: resultSet}, nil
+	case "SELECT @@GLOBAL.SERVER_ID":
+		resultSet, _ := mysql.BuildSimpleTextResultset([]string{"SERVER_ID"}, [][]interface{}{{"10000"}})
```
Comment (teem0n) [inline internal/databases/mysql/binlog_server_handler.go:190]: 10000 should be read from config

File: internal/databases/mysql/binlog_server_handler.go [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+func addRotateEvent(s *replication.BinlogStreamer, pos mysql.Position) error {
+	// create rotate event
+	rotateBinlogEvent := replication.BinlogEvent{}
+	rotateBinlogEvent.RawData = make([]byte, 4+1+4+4+4+2+8)
+	// generate header
+	// timestamp default 4 bytes
+	binlogEventPos := 4
+	rotateBinlogEvent.RawData[binlogEventPos] = byte(replication.ROTATE_EVENT)
+	binlogEventPos++
+	binary.LittleEndian.PutUint32(rotateBinlogEvent.RawData[binlogEventPos:], 10000)
```
Comment (teem0n) [inline internal/databases/mysql/binlog_server_handler.go:37]: 10000 should be read from config

**Review by teem0n — CHANGES_REQUESTED** [review]

File: cmd/mysql/binlog_server.go [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+	"github.com/wal-g/wal-g/internal/databases/mysql"
+	"github.com/wal-g/wal-g/utility"
+)
+
+const (
+	binlogServerShortDescription = "Create server for backup slaves"
+	UntilFlagShortDescr          = "time in RFC3339 for PITR"
+)
+
+var UntilTS string
```
Comment (teem0n) [inline cmd/mysql/binlog_server.go:18]: We should not export this

File: internal/databases/mysql/binlog_server_handler.go [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+	"github.com/wal-g/wal-g/pkg/storages/storage"
+	"github.com/wal-g/wal-g/utility"
+)
+
+var (
+	untilTS      time.Time
+	lastSentGTID string
+)
+
+func HandleEventError(err error, s *replication.BinlogStreamer) {
```
Comment (teem0n) [inline internal/databases/mysql/binlog_server_handler.go:26]: We should not export this

File: internal/databases/mysql/binlog_server_handler.go [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+	tracelog.ErrorLogger.Println("Error during replication", err)
+	ok := s.AddErrorToStreamer(err)
+	for !ok {
+		ok = s.AddErrorToStreamer(err)
+	}
+}
+
+func addRotateEvent(s *replication.BinlogStreamer, pos mysql.Position) error {
+	// create rotate event
+	rotateBinlogEvent := replication.BinlogEvent{}
```
Comment (teem0n) [inline internal/databases/mysql/binlog_server_handler.go:39]: Please add link to event format specs

**Review by teem0n — CHANGES_REQUESTED** [review]

File: internal/databases/mysql/mysql.go [hunk 15]
```diff
@@ -243,6 +254,71 @@ outer:
...
+func provideLogs(folder storage.Folder, dstDir string, startTS, endTS time.Time, p *storage.ObjectProvider) {
+	defer p.Close()
+	_, err := os.Stat(dstDir)
+	if os.IsNotExist(err) {
+		err = os.MkdirAll(dstDir, 0700)
+		if err != nil {
+			ok := p.AddErrorToProvider(err)
+			for !ok {
+				ok = p.AddErrorToProvider(err)
+			}
```
Comment (teem0n) [inline internal/databases/mysql/mysql.go:266]: Let's hide error handling logic, like HandleEventError in binlog_server_handler.go

**Review by teem0n — APPROVED** [review]

**Pull request merged**